datajunction-server/datajunction_server/sql/dag.py

```diff
@@ -1,22 +1,244 @@
+                ),
+            ),
+        ),
+    )
+
+    # Final SELECT statements
+    # ----
+    # If attributes was set to False, we only need to return the dimension nodes
+    if not attributes:
+        return (
+            session.exec(
+                select(Node)
+                .options(joinedload(Node.current))
+                .select_from(paths)
+                .join(Node, paths.c.node_name == Node.name),
+            )
+            .unique()
+            .all()
+        )
+
+    # Otherwise return the dimension attributes, which include both the dimension
+    # attributes on the dimension nodes in the DAG as well as the local dimension
+    # attributes on the initial node
+    group_concat = (
+        func.group_concat
+        if session.bind.dialect.name in ("sqlite",)
+        else func.string_agg
+    )
+    final_query = (
+        select(
+            paths.c.node_name,
+            paths.c.node_display_name,
+            column.name,
+            column.type,
+            group_concat(AttributeType.name, ",").label(
+                "column_attribute_type_name",
+            ),
+            paths.c.join_path,
+        )
+        .select_from(paths)
+        .join(NodeColumns, NodeColumns.node_id == paths.c.node_revision_id)
+        .join(column, NodeColumns.column_id == column.id)
+        .join(ColumnAttribute, column.id == ColumnAttribute.column_id, isouter=True)
+        .join(
+            AttributeType,
+            ColumnAttribute.attribute_type_id == AttributeType.id,
+            isouter=True,
+        )
+        .group_by(
+            paths.c.node_name,
+            paths.c.node_display_name,
+            column.name,
+            column.type,
+            paths.c.join_path,
+        )
+        .union_all(
+            select(
+                NodeRevision.name,
+                NodeRevision.display_name,
+                Column.name,
+                Column.type,
+                group_concat(AttributeType.name, ",").label(
+                    "column_attribute_type_name",
+                ),
+                literal("").label("join_path"),
+            )
+            .select_from(NodeRevision)
+            .join(NodeColumns, NodeColumns.node_id == NodeRevision.id)
+            .join(Column, NodeColumns.column_id == Column.id)
+            .join(
+                ColumnAttribute,
+                Column.id == ColumnAttribute.column_id,
+                isouter=True,
+            )
+            .join(
+                AttributeType,
+                ColumnAttribute.attribute_type_id == AttributeType.id,
+                isouter=True,
+            )
+            .group_by(
+                NodeRevision.name,
+                NodeRevision.display_name,
+                Column.name,
+                Column.type,
+                "join_path",
+            )
+            .where(NodeRevision.id == node_revision.id),
+        )
+    )
+
+    return sorted(
```

Comment:
nit: consider not sorting here. I think most clients will want to sort on their own, instead of depending on the output of some rest api.

Reply:
Agreed, that makes sense! Although this was part of the original code (likely to deal with tests), so I think I'll fix this in a separate PR to keep things a bit cleaner.